### ISSUE TITLE
Built-in terminal

### ISSUE THREAD

**Issue by kdeyko** [kubernetes-sigs/headlamp#2353]
It would be super useful to have a separate section where I could run terminal(s), like Lens [does](https://docs.k8slens.dev/using-lens/terminal/) it. 
As @ckchan91 [notes](https://github.com/unxsist/jet-pilot/issues/24#issuecomment-2371324401):
> In Lens, opening the terminal automatically loads `kubectl` with the corresponding kubeconfig of the cluster being viewed. From the terminal, one can run `kubectl` commands that are not covered by the GUI, such as exporting all the Kubernetes objects etc. This saves the hassle of having to open a terminal and configuring `kubectl` with the correct kubeconfig to interact with the cluster.

It would be beneficial to include a built-in terminal panel at the cluster or context level that:
- opens from the cluster view
- is automatically connected to the currently selected kubeconfig context
- executes kubectl or helm commands using the same credentials as Headlamp
- clearly shows the active context and namespace
- optionally allows multiple terminal tabs
- optionally supports copy/paste and command history

Example UX:
Cluster → Open Terminal → shell opens with the active context pre-selected.

The absence of a terminal creates friction in daily DevOps workflows because users must constantly switch between the Headlamp UI and an external terminal system to run kubectl, helm, and other cluster-scoped commands.

**Comment by joaquimrocha** [comment]
Yes. This is something we definitely want to add.
We have plans to redesign also how the terminal (exec) interaction works. After that, I think we should be able to start working on adding a terminal.

**Comment by DenisBalan** [comment]
@joaquimrocha any updates on the topic?

**Comment by illume** [comment]
@joaquimrocha For your consideration: I think this would be a good one to prioritize. It has a lot of demand in this issue, and apart from above given reasons... it would be very useful for people who want to use headlamp to learn Kubernetes command line (studying for the CKA cert and such). Added to 0.32.0 milestone.

**Comment by joaquimrocha** [comment]
Yes. This should be in our priority list also because we should include kubectl in Headlamp so users don't have to add it or use it externally.

**Comment by illume** [comment]
>We have plans to redesign also how the terminal (exec) interaction works. After that, I think we should be able to start working on adding a terminal.

In the previous release we implemented "activities" which allow persisting terminal views when you're navigating around headlamp, and show tabs (sort of like Visual Studio Code) for switching between them.

I came here after some people on r/kubernetes asked if Headlamp had local terminal yet and wanted to leave an update.

**Comment by siegenthalerroger** [comment]
An interesting extension for the plugin system would be the capability for plugins to provide cli tools, For example the fluxcd cli or cert-manager kubectl plugin would be a boon to have automatically included when the plugin is active.

**Comment by illume** [comment]
@siegenthalerroger good idea. I’ve been thinking something similar for the minikube plugin I’m working on.

**Comment by IraMazurenko** [comment]
It would be beneficial to include a built-in terminal panel at the cluster or context level that:
- opens from the cluster view
- is automatically connected to the currently selected kubeconfig context
- executes kubectl or helm commands using the same credentials as Headlamp
- clearly shows the active context and namespace
- optionally allows multiple terminal tabs
- optionally supports copy/paste and command history

Example UX:
Cluster → Open Terminal → shell opens with the active context pre-selected.

The absence of a terminal creates friction in daily DevOps workflows because users must constantly switch between the Headlamp UI and an external terminal system to run kubectl, helm, and other cluster-scoped commands.

**Comment by illume** [comment]
Thanks @IraMazurenko I added your write up into the original issue.

**Comment by giosueDelgado** [comment]
What is the status of this issue?
